File: app/src/main/java/com/ubergeek42/WeechatAndroid/utils/LineView.java
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

package com.ubergeek42.WeechatAndroid.utils;

import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.text.Spannable;
import android.text.style.ClickableSpan;
import android.text.style.URLSpan;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.ubergeek42.WeechatAndroid.Weechat;
import com.ubergeek42.WeechatAndroid.media.Cache;
import com.ubergeek42.WeechatAndroid.media.Engine;
import com.ubergeek42.WeechatAndroid.media.StrategyUrl;
import com.ubergeek42.WeechatAndroid.relay.Line;
import com.ubergeek42.WeechatAndroid.service.P;
import com.ubergeek42.cats.Kitty;
import com.ubergeek42.cats.Root;

import java.util.List;

import static com.ubergeek42.WeechatAndroid.media.Engine.*;

public class LineView extends View {
    final private @Root Kitty kitty = Kitty.make();

    final private static int THUMBNAIL_AREA_WIDTH = THUMBNAIL_WIDTH + THUMBNAIL_HORIZONTAL_MARGIN * 2;
    final private static int THUMBNAIL_AREA_MIN_HEIGHT = THUMBNAIL_MIN_HEIGHT + THUMBNAIL_VERTICAL_MARGIN * 2;

    private AlphaLayout narrowLayout = null;
    private AlphaLayout wideLayout = null;

    private Spannable text = null;
    private Bitmap image = null;
    private Target target;

    private enum State {TEXT_ONLY, WITH_IMAGE, ANIMATING}
    enum LayoutType {WIDE, NARROW}

    private State state = State.TEXT_ONLY;

    private static int counter = 0;

    public LineView(Context context) {
        this(context, null);
    }

    public LineView(Context context, AttributeSet attrs) {
        super(context, attrs);
        // setLayerType(LAYER_TYPE_HARDWARE, null);
        kitty.setPrefix(String.valueOf(counter++));
    }

    private void reset() {
        text = null;
        wideLayout = narrowLayout = null;
        image = null;
        animatedValue = 0f;
        firstDrawAt = HAVE_NOT_DRAWN;
        state = State.TEXT_ONLY;
        if (animator != null) animator.cancel();
        animator = null;
        Glide.with(getContext()).clear(target);     // will call the listener!
        target = null;
    }

    public void setText(Line line) {
        if (text == line.spannable && getCurrentLayout().getPaint() == P.textPaint) return;
        reset();

        text = line.spannable;

        List<StrategyUrl> candidates = Engine.getPossibleMediaCandidates(getUrls());
        StrategyUrl url = candidates.isEmpty() ? null : candidates.get(0);
        Cache.Info info = url == null ? null : Cache.info(url);

        setLayout(info == Cache.Info.FETCHED_RECENTLY ? LayoutType.NARROW : LayoutType.WIDE);
        invalidate();

        if (url == null || info == Cache.Info.FAILED_RECENTLY) return;

        ensureLayout(LayoutType.NARROW);
        target = Glide.with(getContext())
                .asBitmap()
                .apply(Engine.defaultRequestOptions)
                .listener(Cache.bitmapListener)
                .load(url)
                .into(new Target(THUMBNAIL_WIDTH, getThumbnailHeight()));
    }

    private class Target extends CustomTarget<Bitmap> {
        Target(int width, int height) {
            super(width, height);
        }

        @Override public void onResourceReady(@NonNull Bitmap resource, @Nullable Transition transition) {
            setImage(resource);
        }

        @Override public void onLoadCleared(@Nullable Drawable placeholder) {
            setImage(null);
        }

        // the request seems to be attempted once again on minimizing/restoring the app. to avoid
        // that, clear target soon, but not on current thread as the library doesn't allow it
        @Override public void onLoadFailed(@Nullable Drawable errorDrawable) {
            Target local = target;
            Weechat.runOnMainThread(() -> Glide.with(getContext()).clear(local));
            setImage(null);
        }
    }

    private void setImage(@Nullable Bitmap image) {
        if (this.image == image && !(image == null && state == State.WITH_IMAGE)) return;
        this.image = image;
        if (text == null) return;   // text can be null if called from reset(), in this case don't proceed
        if (shouldAnimateChange()) {
            animateChange();
        } else {
            setLayout(image == null ? LayoutType.WIDE : LayoutType.NARROW);
            invalidate();
        }
    }

    private AlphaLayout getCurrentLayout() {
        return (state == State.WITH_IMAGE) ? narrowLayout : wideLayout;
    }

    private void ensureLayout(LayoutType layoutType) {
        if (layoutType == LayoutType.WIDE && wideLayout == null) wideLayout = AlphaLayout.make(text, P.weaselWidth);
        if (layoutType == LayoutType.NARROW && narrowLayout == null) narrowLayout = AlphaLayout.make(text, P.weaselWidth - THUMBNAIL_AREA_WIDTH);
    }

    private void setLayout(LayoutType layoutType) {
        ensureLayout(layoutType);
        int oldHeight = getCurrentLayout() == null ? -1 : getCurrentLayout().getHeight();
        this.state = layoutType == LayoutType.WIDE ? State.TEXT_ONLY : State.WITH_IMAGE;
        if (getCurrentLayout() != null && oldHeight != getCurrentLayout().getHeight()) requestLayout();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    @Override protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        setMeasuredDimension(P.weaselWidth, getViewHeight(state));
    }

    private int getViewHeight(State state) {
        if (state == State.TEXT_ONLY) return wideLayout.getHeight();
        if (state == State.WITH_IMAGE) return Math.max(narrowLayout.getHeight(), THUMBNAIL_AREA_MIN_HEIGHT);

        int wideHeight = getViewHeight(State.TEXT_ONLY);
        int narrowHeight = getViewHeight(State.WITH_IMAGE);
        return wideHeight + ((int) ((narrowHeight - wideHeight) * animatedValue));
    }

    @Override protected void onDraw(Canvas canvas) {
        if (state == State.ANIMATING) {
            wideLayout.draw(canvas, 1f - animatedValue);
            narrowLayout.draw(canvas, animatedValue);
        } else {
            getCurrentLayout().draw(canvas);
        }
        if (image != null) canvas.drawBitmap(image,
                P.weaselWidth - THUMBNAIL_WIDTH - THUMBNAIL_HORIZONTAL_MARGIN,
                THUMBNAIL_VERTICAL_MARGIN,
                state == State.ANIMATING ? narrowLayout.getAlphaPaint() : null);
        if (firstDrawAt == HAVE_NOT_DRAWN) firstDrawAt = System.currentTimeMillis();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    // see android.text.method.LinkMovementMethod.onTouchEvent
    // todo call performClick()?
    @SuppressLint("ClickableViewAccessibility")
    @Override public boolean onTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_UP) {
            int line = getCurrentLayout().getLineForVertical((int) event.getY());
            int off = getCurrentLayout().getOffsetForHorizontal(line, event.getX());
            ClickableSpan[] links = text.getSpans(off, off, ClickableSpan.class);
            if (links.length > 0) {
                links[0].onClick(this);
                return true;
            }
        }
        return super.onTouchEvent(event);
    }

    public URLSpan[] getUrls() {
        return text.getSpans(0, text.length(), URLSpan.class);
    }

    private int getThumbnailHeight() {
        int height = narrowLayout.getHeight() - THUMBNAIL_VERTICAL_MARGIN * 2;
        if (height < THUMBNAIL_MIN_HEIGHT) height = THUMBNAIL_MIN_HEIGHT;
        if (height > THUMBNAIL_MAX_HEIGHT) height = THUMBNAIL_MAX_HEIGHT;
        return height;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////// animation
    ////////////////////////////////////////////////////////////////////////////////////////////////

    // a value that is used to animate view size and the crossfade. 0f corresponds to the initial
    // text-only layout, 1f to the layout with image. can go both directions
    private float animatedValue = 0f;

    // time in ms when the current text have been first drawn on canvas
    private long firstDrawAt = HAVE_NOT_DRAWN;

    private @Nullable ValueAnimator animator;
    final private static long HAVE_NOT_DRAWN = -1;

    private void animateChange() {
        prepareForAnimation();
        boolean hasImage = image != null;
        boolean needsRelayout = getViewHeight(State.TEXT_ONLY) != getViewHeight(State.WITH_IMAGE);
        float from = hasImage ? 0f : 1f;
        float to = hasImage ? 1f : 0f;
        animator = ValueAnimator.ofFloat(from, to).setDuration(ANIMATION_DURATION);
        animator.addUpdateListener(animation -> {
            animatedValue = (float) animation.getAnimatedValue();
            if (animatedValue == to) {
                state = hasImage ? State.WITH_IMAGE : State.TEXT_ONLY;
                narrowLayout.clearBitmap();
                wideLayout.clearBitmap();
            }
            if (needsRelayout) requestLayout();
            invalidate();
        });
        animator.start();
        state = State.ANIMATING;
    }

    public void cancelAnimation() {
        if (animator != null) animator.cancel();
    }

    // animate layout change—but only if the view is visible and has been visible for some minimum
    // period of time, to avoid too much animation. see https://stackoverflow.com/a/12428154/1449683
    final private Rect _rect = new Rect();
    private boolean shouldAnimateChange() {
        if (!isAttachedToWindow() || getParent() == null)
             return false;
        ((View) getParent()).getHitRect(_rect);
        if (!getLocalVisibleRect(_rect))
            return false;
        long time = System.currentTimeMillis();
        return firstDrawAt != HAVE_NOT_DRAWN && time - firstDrawAt > 50;
    }

    private void prepareForAnimation() {
        ensureLayout(LayoutType.WIDE);
        ensureLayout(LayoutType.NARROW);
        wideLayout.ensureBitmap();
        narrowLayout.ensureBitmap();
    }
}

File: app/src/main/java/com/ubergeek42/WeechatAndroid/views/LineView.kt
package com.ubergeek42.WeechatAndroid.views

import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ClickableSpan
import android.text.style.URLSpan
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.View
import androidx.core.animation.doOnEnd
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.ubergeek42.WeechatAndroid.Weechat
import com.ubergeek42.WeechatAndroid.media.Cache
import com.ubergeek42.WeechatAndroid.media.Config
import com.ubergeek42.WeechatAndroid.media.Engine
import com.ubergeek42.WeechatAndroid.media.Strategy
import com.ubergeek42.WeechatAndroid.media.WAGlideModule
import com.ubergeek42.WeechatAndroid.relay.Line
import com.ubergeek42.WeechatAndroid.upload.f
import com.ubergeek42.WeechatAndroid.upload.i
import com.ubergeek42.WeechatAndroid.upload.main
import com.ubergeek42.WeechatAndroid.utils.invalidatableLazy


private const val ANIMATION_DURATION = 250L // ms


private enum class State(
    val withImage: Boolean,
    val animatingText: Boolean,
    val animatingImage: Boolean,
) {
    TextOnly(false, false, false),                  // wide layout, no image
    TextWithImage(true, false, false),              // narrow layout. image might not be present but expected to be loaded soon
    AnimatingToTextWithImage(true, true, true),     // text only → image
    AnimatingToTextOnly(false, true, false),        // image (without image) → text. runs when the image fails to load
    AnimatingOnlyImage(true, false, true),          // narrow layout; animating only the image
}

class LineView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0,
) : View(context, attrs, defStyle) {
    private val glide = getSafeGlide()

    private var text: Spannable = NoText

    fun setText(line: Line) {
        if (text == line.spannable && currentLayout.usesCurrentPaint()) return

        invalidateInternal()
        setTextInternal(line)
        invalidate()
    }

    private fun invalidateInternal() {
        wideLayoutDelegate.invalidate()
        narrowLayoutDelegate.invalidate()
        text = NoText
        image = null
        firstDrawAt = HAVE_NOT_DRAWN
        state = State.TextOnly
        lastRequestedUrl = null
        animatedValue = 0f
        animator?.cancel()
        animator = null
        glide?.clear(target)    // will call the listener!
        target = null
    }

    private fun setTextInternal(line: Line) {
        this.text = line.spannable

        val (url, cacheInfo) = getUrlInfo(line)
        val prettySureWillShowImage = cacheInfo == Cache.Info.FETCHED_RECENTLY
        state = if (prettySureWillShowImage) State.TextWithImage else State.TextOnly

        maybeRequestLayout()

        if (url != null && cacheInfo != Cache.Info.FAILED_RECENTLY) requestThumbnail(url)
    }

    private fun maybeRequestLayout() {
        if (measuredWidth != wideLayoutWidth || measuredHeight != measureViewHeight())
            requestLayout()
    }

    private var state = State.TextOnly

    val urls: Array<URLSpan> get() = text.getSpans(0, text.length, URLSpan::class.java)

    ////////////////////////////////////////////////////////////////////////////////////////////////

    private val wideLayoutDelegate = invalidatableLazy { AlphaLayout.make(text, wideLayoutWidth) }
    private val narrowLayoutDelegate = invalidatableLazy { AlphaLayout.make(text, narrowLayoutWidth) }

    private val wideLayout by wideLayoutDelegate
    private val narrowLayout by narrowLayoutDelegate

    private val currentLayout get() = when (state) {
        State.TextOnly, State.AnimatingToTextOnly -> wideLayout
        State.TextWithImage, State.AnimatingToTextWithImage, State.AnimatingOnlyImage -> narrowLayout
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////// image
    ////////////////////////////////////////////////////////////////////////////////////////////////

    private var image: Bitmap? = null
    private var target: Target? = null

    private fun getUrlInfo(line: Line): Pair<Strategy.Url?, Cache.Info?> {
        if (Engine.isEnabledAtAll() &&
            Engine.isEnabledForLocation(Engine.Location.CHAT) &&
            Engine.isEnabledForLine(line)) {
            val candidates = Engine.getPossibleMediaCandidates(urls, Strategy.Size.SMALL)
            if (candidates.isNotEmpty()) {
                val url = candidates[0]
                return Pair(url, Cache.info(url))
            }
        }
        return Pair(null, null)
    }

    private var lastRequestedUrl: Strategy.Url? = null
    private fun requestThumbnail(url: Strategy.Url) {
        lastRequestedUrl = url

        glide?.clear(target)

        target = (glide ?: return)
                .asBitmap()
                .apply(Engine.defaultRequestOptions)
                .listener(Cache.bitmapListener)
                .load(url)
                .onlyRetrieveFromCache(Engine.isDisabledForCurrentNetwork())
                .into(Target(Config.thumbnailWidth, measureThumbnailHeight()))
    }

    private inner class Target constructor(width: Int, height: Int) : CustomTarget<Bitmap>(width, height) {
        override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
            setImage(resource)
        }

        // the request seems to be attempted once again on minimizing/restoring the app.
        // to avoid that, clear target soon, but not on current thread--the library doesn't allow it
        override fun onLoadFailed(errorDrawable: Drawable?) {
            target?.let { main { glide?.clear(it) } }
            setImage(null)
        }

        override fun onLoadCleared(placeholder: Drawable?) {
            image = null
        }
    }

    private fun setImage(newImage: Bitmap?) {
        val oldImage = this.image
        if (oldImage == newImage && !(image == null && state == State.TextWithImage)) return
        this.image = newImage

        if (shouldAnimateChange(oldImage, newImage)) {
            animateChange()
        } else {
            state = if (newImage != null) State.TextWithImage else State.TextOnly
            maybeRequestLayout()
            invalidate()
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////// measuring & drawing
    ////////////////////////////////////////////////////////////////////////////////////////////////

    private fun measureViewHeight(): Int {
        fun getTextOnlyHeight() = if (text === NoText) 0 else wideLayout.height

        fun getTextWithImageHeight(): Int {
            val narrowLayoutHeight = if (text === NoText) 0 else narrowLayout.height
            return maxOf(narrowLayoutHeight, Config.thumbnailAreaMinHeight)
        }

        return when (state) {
            State.TextOnly -> getTextOnlyHeight()
            State.TextWithImage, State.AnimatingOnlyImage -> getTextWithImageHeight()
            State.AnimatingToTextOnly, State.AnimatingToTextWithImage -> {
                val textOnlyHeight = getTextOnlyHeight()
                val textWithImageHeight = getTextWithImageHeight()
                textOnlyHeight + ((textWithImageHeight - textOnlyHeight) * animatedValue).i
            }
        }
    }

    private fun measureThumbnailHeight(): Int {
        return (narrowLayout.height - Config.THUMBNAIL_VERTICAL_MARGIN * 2)
                .coerceAtMost(Config.thumbnailMaxHeight)
                .coerceAtLeast(Config.thumbnailMinHeight)
                .coerceAtLeast(1)
    }

    private var wideLayoutWidth: Int = 0
        get() {
            if (field == 0) field = (context as? Activity)?.calculateApproximateWeaselWidth() ?: 1000
            return field
        }

    private val narrowLayoutWidth get() = wideLayoutWidth - Config.thumbnailAreaWidth

    private var oldViewWidth = wideLayoutWidth
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val parentWidth = MeasureSpec.getSize(widthMeasureSpec)
        val viewWidth = if (parentWidth > 0) {
            parentWidth
        } else {
            (context as? Activity)?.calculateApproximateWeaselWidth() ?: 1000
        }

        wideLayoutWidth = viewWidth

        if (oldViewWidth != viewWidth) {
            oldViewWidth = viewWidth
            wideLayoutDelegate.invalidate()
            narrowLayoutDelegate.invalidate()
            if (state.withImage) lastRequestedUrl?.let { requestThumbnail(it) }
        }
        setMeasuredDimension(viewWidth, measureViewHeight())
    }

    override fun onDraw(canvas: Canvas) {
        if (state.animatingText) {
            wideLayout.draw(canvas, 1f - animatedValue)
            narrowLayout.draw(canvas, animatedValue)
        } else {
            currentLayout.draw(canvas)
        }

        image?.let {
            val left = narrowLayoutWidth + Config.THUMBNAIL_HORIZONTAL_MARGIN.f
            val top = Config.THUMBNAIL_VERTICAL_MARGIN.f

            val paint = if (state.animatingImage) {
                            _paint.apply { alpha = (animatedValue * 255).i }
                        } else {
                            null
                        }

            canvas.drawBitmap(it, left, top, paint)
        }

        if (firstDrawAt == HAVE_NOT_DRAWN) firstDrawAt = System.currentTimeMillis()
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////// animation
    ////////////////////////////////////////////////////////////////////////////////////////////////

    // time in ms when the current text have been first drawn on canvas
    private var firstDrawAt = HAVE_NOT_DRAWN

    // a value that is used to animate view size and the crossfade. 0f corresponds to the initial
    // text-only layout, 1f to the layout with image. can go both directions
    private var animatedValue = 0f

    private var animator: ValueAnimator? = null

    // animate layout change—but only if the view is visible
    // and has been visible for some minimum period of time, to avoid too much animation.
    // also don't animate when changing an existing image for a new one due to view resize
    private fun shouldAnimateChange(oldImage: Bitmap?, newImage: Bitmap?): Boolean {
        if (!isAttachedToWindow || parent == null) return false
        if (firstDrawAt == HAVE_NOT_DRAWN) return false
        if (System.currentTimeMillis() - firstDrawAt < 50) return false
        (parent as View).getHitRect(_rect)      // see https://stackoverflow.com/a/12428154/1449683
        if (!getLocalVisibleRect(_rect)) return false
        if (oldImage != null && newImage != null) return false
        return true
    }

    private fun animateChange() {
        val animatingToImage = image != null
        val (from, to) = if (animatingToImage) 0f to 1f else 1f to 0f

        state = if (animatingToImage) {
            if (state == State.TextWithImage) State.AnimatingOnlyImage else State.AnimatingToTextWithImage
        } else {
            State.AnimatingToTextOnly
        }

        animator = ValueAnimator.ofFloat(from, to).also {
            it.duration = ANIMATION_DURATION

            it.addUpdateListener { animation ->
                animatedValue = animation.animatedValue as Float
                maybeRequestLayout()
                invalidate()
            }

            it.doOnEnd {
                state = if (animatingToImage) State.TextWithImage else State.TextOnly
                wideLayoutDelegate.getValueOrNull()?.clearBitmap()
                narrowLayoutDelegate.getValueOrNull()?.clearBitmap()
                animator = null
            }
        }

        animator?.start()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        animator?.end()
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////// touch
    ////////////////////////////////////////////////////////////////////////////////////////////////

    // AS suggests calling performClick(), which does stuff like playing tap sound and accessibility,
    // but LinkMovementMethod doesn't seem to be calling any methods like that, so we are not doing
    // it either. on long click, we call performLongClick(), which also does haptic feedback.
    // we are consuming all touch events mostly because it works well. perhaps only handle up/down?
    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        gestureDetector.onTouchEvent(event)
        return true
    }

    private val gestureDetector = GestureDetector(Weechat.applicationContext,
            object : SimpleOnGestureListener() {
                override fun onLongPress(event: MotionEvent) {
                    performLongClick()
                }

                override fun onDoubleTap(event: MotionEvent): Boolean {
                    onDoubleTapListener?.invoke()
                    return true
                }

                // see android.text.method.LinkMovementMethod.onTouchEvent
                override fun onSingleTapUp(event: MotionEvent): Boolean {
                    val currentLayout = this@LineView.currentLayout
                    val line = currentLayout.getLineForVertical(event.y.i)

                    if (event.x in currentLayout.getHorizontalTextCoordinatesForLine(line)) {
                        val offset = currentLayout.getOffsetForHorizontal(line, event.x)
                        val links = text.getSpans(offset, offset, ClickableSpan::class.java)

                        if (links.isNotEmpty()) {
                            links.first().onClick(this@LineView)
                            return true
                        }
                    }

                    return false
                }
            })

    var onDoubleTapListener: (() -> Unit)? = null
}


// these two are only used in single methods on the main thread
@Suppress("ObjectPropertyName") private val _paint = Paint()
@Suppress("ObjectPropertyName") private val _rect = Rect()

const val HAVE_NOT_DRAWN = -1L

private val NoText = SpannableString("error")   // just so that we don't need to say !!


fun View.getSafeGlide() = if (WAGlideModule.isContextValidForGlide(context)) {
                               Glide.with(context)
                           } else {
                               null
                           }
